src/coprocessor/dag/batch_executor/statistics.rs

```diff
@@ -24,17 +29,155 @@ pub struct BatchExecuteStatistics {
 }
 
 impl BatchExecuteStatistics {
-    pub fn new(ranges: usize) -> Self {
+    pub fn new(executors: usize, ranges: usize) -> Self {
+        let mut summary_per_executor = Vec::new();
+        summary_per_executor.resize_with(executors, || None);
+
         Self {
+            summary_per_executor,
             scanned_rows_per_range: vec![0; ranges],
             cf_stats: crate::storage::Statistics::default(),
         }
     }
 
     pub fn clear(&mut self) {
+        for item in self.summary_per_executor.iter_mut() {
+            *item = None;
+        }
         for item in self.scanned_rows_per_range.iter_mut() {
             *item = 0;
         }
         self.cf_stats = crate::storage::Statistics::default();
     }
 }
+
+/// Execution summaries to support `EXPLAIN ANALYZE` statements.
+#[derive(Default)]
+pub struct ExecSummary {
+    /// Total time cost in this executor.
+    pub time_processed_ms: Arc<atomic::AtomicUsize>,
+
+    /// How many rows this executor produced totally.
+    pub num_produced_rows: usize,
+
+    /// How many times executor's `next()` is called.
+    pub num_iterations: usize,
+}
+
+impl ExecSummary {
+    #[inline]
+    pub fn merge_into(self, target: &mut Self) {
+        target.time_processed_ms.fetch_add(
+            self.time_processed_ms.load(atomic::Ordering::Relaxed),
+            atomic::Ordering::Relaxed,
+        );
+        target.num_produced_rows += self.num_produced_rows;
+        target.num_iterations += self.num_iterations;
+    }
+}
+
+pub trait ExecSummaryCollector: Send {
+    type TimeRecorderGuard;
+
+    /// Creates a new instance with specified output slot index.
+    fn new(output_index: usize) -> Self
+    where
+        Self: Sized;
+
+    /// Creates a guard that will collect elapsed time as the scope duration when it is dropped.
+    fn collect_scope_duration(&self) -> Self::TimeRecorderGuard;
+
+    /// Increases produced rows counter.
+    fn inc_produced_rows(&mut self, rows: usize);
+
+    /// Increases iterations counter.
+    fn inc_iterations(&mut self);
+
+    /// Takes and appends current execution summary into `target`.
+    fn collect_into(&mut self, target: &mut [Option<ExecSummary>]);
+}
+
+/// A helper to record duration for timing fields for `ExecSummary` based on `Drop`.
+pub struct ExecSummaryTimeGuard {
+    start_time: crate::util::time::Instant,
+    collect_target: Arc<atomic::AtomicUsize>,
+}
+
+impl Drop for ExecSummaryTimeGuard {
+    #[inline]
+    fn drop(&mut self) {
+        let elapsed = (self.start_time.elapsed_secs() * 1000f64) as usize;
+        self.collect_target
+            .fetch_add(elapsed, atomic::Ordering::Relaxed);
+    }
+}
+
+/// A normal `ExecSummaryCollector`. Acts like `collect = true`.
+pub struct ExecSummaryCollectorNormal {
```

Comment:
How about just using a boolean to decide whether we need `ExecSummaryCollector`? Or maybe use an `Option`?

Reply:
Boolean or Option will introduce runtime cost (to check the bool), while using this way there will be completely no cost when it is not needed.

Reply:
How about `ExecSummaryCollectorEnabled`?

Reply:
That's ok